### PR TITLE
Restructure benchmarks, add Hessian test cases

### PR DESCRIPTION
**Pull request by gdalle**
- Restructure Jacobian benchmarks to make them more modular
- Add Hessian benchmarks based on the JuMP hessian article (https://www.tandfonline.com/doi/full/10.1080/10556788.2018.1480625)

> This PR renames benchmark groups and changes the nesting, so it cannot be compared with the `main` branch. But that's ok cause I didn't change the `src` folder, so performance stays the same

### CODE REVIEW

File: benchmark/benchmarks.jl [hunk 1]
```diff
@@ -1,74 +1,19 @@
...
-    for N in (28, 128)
-        J_global["conv"]["size=$(N)x$(N)x3"][setname] = @benchmarkable jacobian_sparsity(
-            $LAYER, $(rand(N, N, 3, 1)), $method
-        )
-    end
-    # TODO: benchmark local sparsity tracers on LeNet-5 CNN
+for G in SET_TYPES
+    H = Set{Tuple{Int,Int}}
+    SUITE["Jacobian"][nameof(G)] = jacbench(TracerSparsityDetector(G))
+    SUITE["Hessian"][(nameof(G), nameof(H))] = hessbench(TracerSparsityDetector(G, H))
```
Comment (adrhill) [inline benchmark/benchmarks.jl:16]: Can we benchmark `TracerLocalSparsityDetector` as well?

Comment (gdalle) [reply benchmark/benchmarks.jl]: done in the last commit

**Comment by adrhill** [comment]
Looks like the issue lies in my best friend `similar`:
```Julia
Closest candidates are:
  primal(!Matched::SparseConnectivityTracer.Dual)
   @ SparseConnectivityTracer ~/work/SparseConnectivityTracer.jl/SparseConnectivityTracer.jl/src/tracers.jl:217

Stacktrace:
  [1] _broadcast_getindex_evalf
    @ ./broadcast.jl:709 [inlined]
  [2] _broadcast_getindex
    @ ./broadcast.jl:682 [inlined]
  [3] getindex
    @ ./broadcast.jl:636 [inlined]
  [4] copy
    @ ./broadcast.jl:942 [inlined]
  [5] materialize
    @ ./broadcast.jl:903 [inlined]
  [6] similar(a::Array{Float64, 3}, ::Type{SparseConnectivityTracer.Dual{Float64, SparseConnectivityTracer.GradientTracer{BitSet}}}, dims::Tuple{Int64, Int64, Int64})
```

**Comment by adrhill** [comment]
Can you remove the hotfix on `primal`? #106 should fix the actual issue on `similar`.

**Comment by adrhill** [comment]
Feel free to merge if #106 fixes the issue!

**Review by adrhill — APPROVED** [review]

**Comment by adrhill** [comment]
#106 has been merged.

File: src/tracers.jl [hunk 2]
```diff
@@ -215,6 +215,7 @@ Dual(primal::P, tracer::T) where {P,T} = Dual{P,T}(primal, tracer)
 # TODO: support ConnectivityTracer
 
 primal(d::Dual) = d.primal
+primal(p::Real) = p  # TODO: should this be needed?
```
[the remark this answers is not in the document]
Comment (gdalle) [reply src/tracers.jl]: removed

**Pull request merged**